[RISCV] tt-ascalon-d8 vector scheduling

ppenzin opened the pull request:
Add the vector scheduling model for tt-ascalon-d8 and corresponding llvm-mca tests.

ppenzin committed:
[WIP][RISCV] tt-ascalon-d8 vector scheduling

Drive-by: additional tuning knobs.

Partial scheduling model for vector operations.

ppenzin commented:
This is far from complete, in addition to more latencies it also needs tests.

I have some questions. Does max(ReleaseAtCycles) have to be less than Latency? Can AcquireAtCycles overlap for different pipelines?

mshockwave reviewed (commented):
could you also add llvm-mca tests?

ppenzin committed:
Simplify release/acquire

Out of order cores don't necessarily need to chain acquire-at and
release-at cycles.

ppenzin committed:
Add vector mask and permute operations

ppenzin committed:
Add read advances and a few more instructions

Add as number of read advances, VExtV, and VRGatherVV cost function.

ppenzin committed:
Vector memory ops

ppenzin committed:
Update llvm/lib/Target/RISCV/RISCVSchedTTAscalonD8.td

Co-authored-by: Craig Topper <craig.topper@sifive.com>

ppenzin committed:
Clean up copied latencies, remove TODOs

ppenzin commented on llvm/lib/Target/RISCV/RISCVSchedTTAscalonD8.td:
```diff
@@ -8,19 +8,85 @@
+}
+
+/// Linear LMUL scaling starting from smallest fractional LMUL
+class AscalonGetCyclesLMULFractional<string mx, int base> {
```
Not anymore

ppenzin committed:
Clear instances of cycles-1=0, add tests

ppenzin commented:
gentle ping

ppenzin commented:
> yeah speaking of this, we really need a more elegant way to clean up -- potentially centralize -- all scheduling model tests. Otherwise it's really easy to miss test coverage.

Yeah, we do. There was discussion at the summit about potentially using UnifiedDB for this, thought it's support for vector constraints isn't complete.

ppenzin commented on llvm/lib/Target/RISCV/RISCVSchedTTAscalonD8.td:
```diff
@@ -316,10 +409,634 @@ def : ReadAdvance<ReadSHXADD32, 0>;
+    defm "" : LMULWriteResMX<"WriteVSTOX8", [AscalonLS], mx, IsWorstCase>;
+  }
+}
+foreach mx = ["MF4", "MF2", "M1", "M2", "M4", "M8"] in {
```
Yes, though I am a bit hesitant changing this for two reasons: `SchedMxListF` has a comment describing it as floating point, and there are further sections that have a variation of this list (next would be missing "MF4", then "MF2", etc). I don't feel very strongly either way though.

ppenzin committed:
Review feedback

Remove redundant latencies of 1. Use `-instruction-tables=full` in
llvm-mca test command lines.

ppenzin commented on llvm/lib/Target/RISCV/RISCVSchedTTAscalonD8.td:
```diff
@@ -316,10 +409,634 @@ def : ReadAdvance<ReadSHXADD32, 0>;
+    defm "" : LMULWriteResMX<"WriteVLDE",    [AscalonLS], mx, IsWorstCase>;
+    defm "" : LMULWriteResMX<"WriteVLDFF",   [AscalonLS], mx, IsWorstCase>;
+  }
+  let Latency = 1 in
```
Addressed

ppenzin commented on llvm/lib/Target/RISCV/RISCVSchedTTAscalonD8.td:
```diff
@@ -316,10 +409,634 @@ def : ReadAdvance<ReadSHXADD32, 0>;
+def : WriteRes<WriteRdVLENB, [AscalonFXA]>;
+
+// Configuration-Setting Instructions
+let Latency = 1 in {
```
Addressed

ppenzin committed:
Move tune settings out

mshockwave reviewed (approved):
LGTM w/ one minor comment

ppenzin commented on llvm/lib/Target/RISCV/RISCVSchedTTAscalonD8.td:
```diff
@@ -316,10 +409,625 @@ def : ReadAdvance<ReadSHXADD32, 0>;
+  }
+}
+
+// VLD*R is LMUL aware
```
yes, I think it was supposed to mean 'originally LMUL-aware'

ppenzin committed:
Clarify LMUL awareness for some ops
